CreateDB

seiya-takeda opened the pull request:
# what
DB設計・実装のため

# why
サービスに必須の機能のため

seiya-takeda committed:
CreateDB

yuki-uchida commented on chat-space/README.md:
```diff
@@ -22,3 +22,45 @@ Things you may want to cover:
+
+|Column|Type|Options|
+|------|----|-------|
+|user_id|integer|null: false, foreign_key: true|
```
外部キー制約をつけたい場合にはintegerではなくrefferenceと書いておきましょう

seiya-takeda replied:
記載の仕方としては下記で相違ないでしょうか？
|user_id|reffrence|null: false, foreign_key: true|

seiya-takeda replied:
"references"かと思ったのですが.....

yuki-uchida replied:
references ですね！失礼いたしました

yuki-uchida replied:
記載のフォーマットはそちらで大丈夫です

seiya-takeda replied:
承知致しました、ご確認頂きありがとうございます。

yuki-uchida replied:
すいません、変更が確認できません

yuki-uchida commented on chat-space/README.md:
```diff
@@ -22,3 +22,45 @@ Things you may want to cover:
+|tweet_id|integer|null: false, foreign_key: true|
+|user_id|integer|null: false, foreign_key: true|
+|image|text|null: false, foreign_key: true|
+|tweet|text|null: false, foreign_key: true|
```
画像がテキストのどちらかがあれば投稿できるようにしましょう

seiya-takeda replied:
現在は画像とテキストの両方がないと投稿できないようになっている、ということでしょうか？

yuki-uchida replied:
そうですね、imageもtweetもnull:falseになっていますので両方ないと投稿できない状況です

seiya-takeda replied:
imageもtextもnull:falseを削除してしまうと、何も投稿内容がない状態でも投稿できるようになってしまうと思います。
"imageもしくはtextいずれかがnullでなければ投稿可能と設定"と考えた調べてみたのですが解決法が分からず。。。
・そもそも考え方が違うでしょうか？
・上記を設定するための書き方があるのでしょうか？
ご確認をお願いします。

yuki-uchida replied:
こちらはモデルのvalidationで設定するのが一般的です。
片方あれば投稿できる状態を作るためには、DB側には何も制約をつけず、モデル側でvalidationをするのが良いでしょう。

seiya-takeda replied:
ありがとうございます、validation確認致しました。
validationでpresenceを用い特定のカラムが空でないことをチェックすることは理解しました。
ただ最後の記載方法が調べたのですが分からず。。。下記のような記載はできるのでしょうか？
・validates : (image or textのいずれか), presence: true

下記ではまたimageとtext 両方にデータがないと投稿できなくなると認識しております。
・validates : image, presence: true
　validates : text, presence: true

ご確認をお願いします。

yuki-uchida replied:
このように、複数カラムにまたがったvalidationをしたい場合には、カスタムvalidationを作ると良いでしょう
https://qiita.com/n-oshiro/items/4a3188be66dd0e18bae5

seiya-takeda replied:
ご教示頂きありがとうございます。お送り頂きました記事を確認させていただきました。
記事の内容について、そもそも理解が正しいのかの認識を合わせたく
現状下記のように理解しているのですがご確認を頂きたく思います。
（1~4はリンク先の記事を参照しています。）

1.appディレクトリ内にvalidatorsディレクトリを作成する

2. カスタムバリデータを作成し、app/validators ディレクトリに保存します。
 ① def validate_each(record, attribute, value)
　　→()内に跨るカラムを記載する
 ②unless value =~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
　　→”~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/I”を入力されないという条件設定
　　→”~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/I”がカラムに入力されたらエラー文”attributeを返す”
　　→この認識は正しいでしょうか？
　　⇒今回の場合は特に指定が無いため、3行目（unless~endまで）は記載しなくてよいのでしょうか？

3.カスタムバリデータを有効するためconfig/application.rbを修正します
　　⇢自作したカスタムバリデータを自動で読み込むために下記一文を追記するという理解でよろしいでしょうか？
　　    config.autoload_paths += Dir["#{config.root}/app/validators"]

4.app/model.tweetへ今回のケースですと、（一番最初にTweetValidatorを定義していること前提に）
　　　validates :tweet, tweet: true # <- カスタムバリデータ(TweetValidator)
 のようなコードを追記するということでしょうか

kuroko918 replied:
そうですね、該当のパラメータがある時のみvalidationをかける場合は、validates :tweet, tweet: trueとして頂くと大丈夫です

kuroko918 reviewed (commented):
バリデーションについて詳しく調べて頂いて素晴らしいですが、このブランチはDB設計に関するブランチかと思いますので、valodationなどmodel作成はこのDB設計のブランチが完成して、別のブランチで実装された方が良いです。
また、修正が反映されていない様なので、確認お願いします。

kuroko918 commented:
ですので、今回はREADMEのみの修正をgithub上にあげて頂き、他のファイルは一旦github上から削除して下さい。

seiya-takeda commented:
ご確認頂きありがとうござます。
・DB設計の際に新たなブランチを作成し、別のブランチで管理します。
　⇢今回のReadMeブランチでは割愛します。
・README.mdを編集しました。以前ご指摘頂いていた45~46行目はモデル側で設計するため、
両方ともnull: falseを削除しております。
・>他のファイルは一旦github上から削除して下さい。
　　⇢アプリケーションchat-space以下に格納されいるREADME.mdファイル以外すべて削除するということでしょうか？

miyagit commented on chat-space/README.md:
```diff
@@ -22,3 +22,45 @@ Things you may want to cover:
+|Column|Type|Options|
+|------|----|-------|
+|user_id|integer|null: false, foreign_key: true|
+|user_name|string|null: false, foreign_key: true|
```
こちらnameで大丈夫ですね。

seiya-takeda replied:
承知しました。user_name⇢nameに変更します

miyagit commented on chat-space/README.md:
```diff
@@ -22,3 +22,45 @@ Things you may want to cover:
+|------|----|-------|
+|user_id|integer|null: false, foreign_key: true|
+|user_name|string|null: false, foreign_key: true|
+|login_id|string|null: false, foreign_key: true|
```
login_idとは何でしょうか？

seiya-takeda replied:
login_passwordのミスですね、失礼しました。

miyagit replied:
こちら削除してますか？？

seiya-takeda replied:
こちらミスで本来ログインする際のパスワードを管理するlogin_passwordが"正"です。
変更済かと思いますが何か操作等間違っているのでしょうか。。。
https://gyazo.com/192ca95a0e9857e692856f691d30273f

seiya-takeda replied:
> プルリク提出していただいたのですが、pushできていますでしょうか？

syncを押せばpushできているという認識なのですが正しいでしょうか？
念の為再度syncしましたのでご確認をお願いします。

miyagit commented on chat-space/README.md:
```diff
@@ -22,3 +22,45 @@ Things you may want to cover:
+
+|Column|Type|Options|
+|------|----|-------|
+|tweet_id|integer|null: false, foreign_key: true|
```
ここはただのidで良いですかね。

seiya-takeda replied:
承知しました、変更済です

shingoteshima commented:
プルリク提出していただいたのですが、pushできていますでしょうか？

yuki-uchida reviewed (changes requested):
pushが出来てなさそうです

yuki-uchida commented:
syncを押せばpushできているという認識なのですが正しいでしょうか？
念の為再度syncしましたのでご確認をお願いします。

その認識であっています。しかし、現状確認が変更出来ていないので、ブランチが違うということは無いでしょうか？

seiya-takeda committed:
Correction

メンターからのFBを受け
・user_idカラム型をinteger⇢referencesへ変更
・tweet_id⇢idにリネーム
・imageカラムとtweetカラムからnull:false,を削除
・user_name⇢nameへリネーム
・login_id⇢login_passwordへリネーム

seiya-takeda commented:
・ブランチはReadMeブランチで行っております。
https://gyazo.com/012645eb289fcbf8ebf6dbbfd5e16d88

・また下記ご回答をお願いします。
>>他のファイルは一旦github上から削除して下さい。
　>⇢アプリケーションchat-space以下に格納されいるREADME.mdファイル以外すべて削除するということでしょうか？

yuki-uchida commented:
たった今変更が確認できました。


プルリクというものは、そのブランチにおいて変更を加えた差分についてレビューを頂くものなのですが、今回のプルリクは、DBの設計図を考えるものなので、実際にDBを作ったりモデルを作ったりするのはまだ早いです。そのため、今回のプルリクに含めて欲しいのは、DBの設計図のみ(Readme)となります。

ファイルを消して欲しいという訳ではなく、プルリクの範囲外で変更を加えたものをプルリクに含めないで欲しい
ということです。

このページのFile changedのタブをクリックして頂くと、変更を加えたファイルが全て出てきますが、Readme以外のファイルは本プルリクにおいては必要ないので、そのファイルはコミットを取り消して、Readmeのみが残るようにしましょう

seiya-takeda commented:
ご返信頂いた内容に関しましては理解致しました。
ReadMe以外のファイルのみコミットを取り消すことはできるのでしょうか？
試してみたのですが、ReadMeも同時にプルリクから外れてしまいます。。。

TOMARUMARU commented:
コミットを全て取り消して、ReadMeのみ再度コミットし、強制pushすればいいと思います

それか一旦ReadMe以外のファイルは消して、その消した差分をpushするとかですね

seiya-takeda commented:
操作としては、git reset --hard HEAD^でコミット前に戻し、ReadMeのみ選択肢pushするということでしょうか？

TOMARUMARU commented:
そうですね、ただhardだと変更差分も消えてしまうのでsoftにしましょう

seiya-takeda commented:
こちら試してみたのですが何も変化がなく（下記URLご参照ください）解決できないので一旦ReadMEブランチを削除し新たにブランチを作成し、改めてプルリクしようと思うのですが大丈夫でしょうか？
https://gyazo.com/04577ed2cde13a252400ea71510e30d7
